### PR TITLE
Poll for new notifications

### PR DESCRIPTION
**Pull request by samgst-amazon**
Implement polling and local caching logic for remote notification json file. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description

Start an app level service that runs on startup that polls for new notifications once at startup and every 10 minutes after. Notifies listeners to process notifications when appropriate

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import java.time.Duration
+
+private const val NOTIFICATION_ENDPOINT = "https://idetoolkits-hostedfiles.amazonaws.com/Notifications/JetBrains/1.json" // TODO: Replace with actual endpoint
+private const val MAX_RETRIES = 3
+private const val RETRY_DELAY_MS = 1000L
+
+interface NotificationPollingService {
+    fun startPolling()
+    fun addObserver(observer: (Path) -> Unit)
+    fun dispose()
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:33]: This function is not required in the interface

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import java.nio.file.Paths
+import java.time.Duration
+
+private const val NOTIFICATION_ENDPOINT = "https://idetoolkits-hostedfiles.amazonaws.com/Notifications/JetBrains/1.json" // TODO: Replace with actual endpoint
+private const val MAX_RETRIES = 3
+private const val RETRY_DELAY_MS = 1000L
+
+interface NotificationPollingService {
+    fun startPolling()
+    fun addObserver(observer: (Path) -> Unit)
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:32]: The path should be a constant, we probably dont need to send it here

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+import java.io.InputStream
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.time.Duration
+
+private const val NOTIFICATION_ENDPOINT = "https://idetoolkits-hostedfiles.amazonaws.com/Notifications/JetBrains/1.json" // TODO: Replace with actual endpoint
+private const val MAX_RETRIES = 3
+private const val RETRY_DELAY_MS = 1000L
+
+interface NotificationPollingService {
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:30]: Is an interface required?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        } catch (e: Exception) {
+            false
+        }
+    }
+}
+
+@Service(Service.Level.APP)
+class NotificationPollingServiceImpl :
+    NotificationPollingService,
+    PersistentStateComponent<NotificationPollingServiceImpl.State>,
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:50]: We should probably have a separate persistent state class that keeps track of the etag and get its instance here

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    private fun pollForNotifications(): Boolean {
+        var retryCount = 0
+        var lastException: Exception? = null
+
+        while (retryCount < MAX_RETRIES) {
+            try {
+                val newETag = getNotificationETag()
+                if (newETag == NotificationState.getState().etag) {
+                    RunOnceUtil.runOnceForApp(this::class.qualifiedName.toString()) {
+                        // try startup notifications regardless of file change
```
[the remark this answers is not in the document]
Comment (manodnyab) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:89]: Would it make sense to use an atomic boolean here to check this(See ToolWindowStartupActivity)

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                request.connection.headerFields["ETag"]?.firstOrNull() ?: ""
+            }
+
+    private fun emitFailureMetric(exception: Exception?) {
+        // todo: add metric
+    }
+
+    fun addObserver(observer: (Unit) -> Unit) {
+        observers.add(observer)
+    }
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:127]:
```suggestion
    fun addObserver(observer: () -> Unit) = observers.add(observer)
   
```

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import com.intellij.ide.util.RunOnceUtil
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.startup.ProjectActivity
+
+internal class NotificationServiceInitializer : ProjectActivity {
+
+    override suspend fun execute(project: Project) {
+        val service = NotificationPollingService.getInstance()
+        ProcessNotificationsBase()
+        RunOnceUtil.runOnceForApp(this::class.qualifiedName.toString()) {
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt:15]: Do we need this? Will this startup activity anyway run only once?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+import com.intellij.ide.util.RunOnceUtil
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.startup.ProjectActivity
+
+internal class NotificationServiceInitializer : ProjectActivity {
+
+    override suspend fun execute(project: Project) {
+        val service = NotificationPollingService.getInstance()
+        ProcessNotificationsBase()
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt:14]: why is this required?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt [hunk 9]
```diff
@@ -3,15 +3,34 @@
...
+    val mapper = jacksonObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+}
 
 class ProcessNotificationsBase {
+
     init {
-        // TODO: install a listener for the polling class
+        NotificationPollingService.getInstance().addObserver {
+            val list = getNotificationsFromFile()
+            println("here")
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt:23]: nit: can be removed

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt (hunk 8 above)
Comment (samgst-amazon) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt]: ProcessNotificationsBase needs to initialize somehow (after the pollingservice init and before the start polling begins) so that it's observer is able to be notified on startup. I tried some other ways of implementing, but this is what I found to work

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt (hunk 7 above)
Comment (samgst-amazon) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt]: Switched to Atomic Boolean, but the same sort of logic applies. Because StartupActivity is obsolete in favor of ProjectActivity, we need to ensure we are only creating one instance of the polling service per IDE session. Otherwise, when a user opens multiple projects, it would start multiple instances of the polling service

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    override fun canBeParsed(data: InputStream): Boolean =
+        try {
+            NotificationMapperUtil.mapper.readValue<NotificationsList>(data)
+            true
+        } catch (e: Exception) {
+            false
+        }
+}
+
+@State(name = "NotificationETagState", storages = [Storage("aws.xml", roamingType = RoamingType.DISABLED)])
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:43]: nit: do we need to specify the roaming type?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+object ETagState : PersistentStateComponent<ETagState.State> {
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:44]: Can we convert this to a class with a getInstance method?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    data class State(
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:45]: This return type could lie outside this class.
nit: maybe add a more descriptive name.
eg reference class: EcsExecCommandSettings

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    private fun pollForNotifications(): Boolean {
+        var retryCount = 0
+        var lastException: Exception? = null
+
+        while (retryCount < MAX_RETRIES) {
+            try {
+                val newETag = getNotificationETag()
+                if (newETag == ETagState.getState().etag) {
+                    if (firstPollDone.compareAndSet(false, true)) {
+                        notifyObservers()
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:95]: nit: maybe we could add a comment on why this is required

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+                    Thread.sleep(backoffDelay)
+                }
+            }
+        }
+        emitFailureMetric(lastException)
+        return false
+    }
+
+    private fun getNotificationETag(): String =
+        HttpRequests.request(NOTIFICATION_ENDPOINT)
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:120]: Should this be under a try catch block?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+                    .get()
+                ETagState.getState().etag = newETag
+                return true
+            } catch (e: Exception) {
+                lastException = e
+                LOG.error(e) { "Failed to poll for notifications (attempt ${retryCount + 1}/$MAX_RETRIES)" }
+                retryCount++
+                if (retryCount < MAX_RETRIES) {
+                    val backoffDelay = RETRY_DELAY_MS * (1L shl (retryCount - 1))
+                    Thread.sleep(backoffDelay)
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:111]: is it possible to use delay instead of thread.sleep?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+data class NotificationEtagConfiguration(
+    var etag: String? = null,
+)
+
+@Service(Service.Level.APP)
+internal final class NotificationPollingService : Disposable {
+    private val isFirstPoll = AtomicBoolean(true)
+    private val observers = mutableListOf<() -> Unit>()
+    private val alarm = AlarmFactory.getInstance().create(Alarm.ThreadToUse.POOLED_THREAD, this)
+    private val scope = CoroutineScope(getCoroutineBgContext())
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPollingService.kt:97]: `CoroutineScope` should always be injected by platform

**Review by rli — APPROVED** [review]

File: plugins/core/core/src/software/aws/toolkits/core/utils/RemoteResourceResolver.kt [hunk 17]
```diff
@@ -32,9 +32,11 @@ class DefaultRemoteResourceResolver(
     private fun internalResolve(resource: RemoteResource): Path {
         val expectedLocation = cacheBasePath.resolve(resource.name)
         val current = expectedLocation.existsOrNull()
-        if (current != null && !isExpired(current, resource)) {
-            LOG.debug { "Existing file ($current) for ${resource.name} is present and not expired - using it." }
-            return current
+        if (resource.name != "notifications.json") {
```
Comment (samgst-amazon) [inline plugins/core/core/src/software/aws/toolkits/core/utils/RemoteResourceResolver.kt:35]: @rli This feels hacky, but its one way to get around this issue, the other way I think we could maybe do this is to just set the expiry time to a very short window so that the file is always seen as expired by the time polling makes this call to the endpoint
